### DISCUSSION TITLE
M5Stack Atom S3?

### DISCUSSION THREAD

**Discussion by tino** [micropython/micropython#12746] [ESP32]
There is a version for the [M5Stack Atom](https://micropython.org/download/?vendor=M5%20Stack), but what about one for the Atom S3? It contains an LCD screen instead of a led matrix on the normal Atom. 

Is there a way I can set this up myself? Create a pull request? How would I even go about discovering what's needed?

**Comment by mattytrentini** [comment]
I have been running the `ESP32_GENERIC_S3` build on the AtomS3 but have wanted to create a dedicated board definition for it. 

Creating board definitions - as a general topic - would benefit with documentation (or at least a wiki page?). Some notes specific to the AtomS3 board, if you wanted to have a go:

* Base it on `ESP32_GENERIC_S3`
  * The default, quad SPI variant, not `SPIRAM_OCT`, which can be removed
  * 8MB flash is configured, which is correct
* [Awesome MicroPython](https://awesome-micropython.com/) has drivers for the MPU and display
  * They'll need to be merged to micropython-lib if we want to make them part of the board definition
  * But we could at least add a README to explain how to install/use them
* `pins.csv` should be configured since there are only a relatively small number available
  * It allows a subset of pins to be available - and named alias' to be defined (accessible from `Pin.board.[PINNAME]`)
  * This is a relatively new feature for the ESP32 port and hasn't been widely used yet
  * Maybe see [UM_TINYS3/pins.csv](https://github.com/micropython/micropython/blob/master/ports/esp32/boards/UM_TINYS3/pins.csv) for an example

And yes, clone the MicroPython repo, take a branch, and create your own `ports/esp32/boards/M5STACK_ATOMS3` folder with the appropriate files. Raise a PR for the branch when you're done.

If you think that's something you want to try, I'll test and review the PR! Otherwise I'll tackle it when I get a chance - and ask _you_ to review it. :)

**Reply by rdlaner** [reply]
@mattytrentini I just got an AtomS3 as well and will be looking to run MicroPython on it when I can find the time 😬. Regarding the display driver, is this the one you were indicating was in Awesome MicroPython?
https://github.com/russhughes/s3lcd

**Reply by mattytrentini** [reply]
That's an excellent driver but is written in C meaning it's necessary to build the code into firmware. Currently it's not possible to easily include a C display driver, even if it's in micropython-lib. 

I was actually expecting to find a good ST7789 Python driver - but all the options I've looked at are a little lacking. Still, could at least make a Python driver easy to install with mip...

**Reply by russhughes** [reply]
@mattytrentini what are you looking for in a python ST7789 driver that is missing?

**Reply by mattytrentini** [reply]
Mostly to be based on framebuf so general drawing operations are relatively fast. In fact, I want to measure a framebuf-based driver to see what the performance is like, particularly compared to your C-based drivers. It's not clear to me why a C-based driver will be much faster (for operations that are supported in framebuf)...

**Reply by arizzi** [reply]
did any of you manage to use the AtomS3 display with pure micropython (not uiflow) even with a slow driver? it seems the python implementations from Awesome micropython do not support the AtomS3 resolution

**Reply by russhughes** [reply]
@mattytrentini @arizzi I'll have some time this weekend to look at this.

**Reply by arizzi** [reply]
btw for the accelerometer this one works https://github.com/tuupola/micropython-mpu6886 with SoftI2C  on 39,38
```
from mpu6886 import MPU6886
from machine import SoftI2C

i2cs = SoftI2C(scl=Pin(39), sda=Pin(38))
accsensor = MPU6886(i2cs)

```

**Reply by arizzi** [reply]
it seems just adding the 128x128 mode where needed here https://github.com/russhughes/st7789py_mpy/blob/master/lib/st7789py.py
it works with something like:

```
tft = ST7789(
        spi,
        128,
        128,
        reset=Pin(34, Pin.OUT),
        cs=Pin(15, Pin.OUT),
        dc=Pin(33, Pin.OUT),
        backlight=Pin(16, Pin.OUT),
        rotation=0)
```

**Reply by arizzi** [reply]
forked with the 128x128 addition here, if you want I can make a PR https://github.com/arizzi/st7789py_mpy

**Reply by russhughes** [reply]
@arizzi I updated the driver to work with the AtomS3. I Added color_order, x,y offsets and a couple of examples.
